### DISCUSSION TITLE
Best way to work with Symfony and SPAs

### DISCUSSION THREAD

**Discussion by Nek-** [symfony/symfony#42847] [Q&A] [answered]
Hello folks,

I tried to make a clean development infrastructure of a project that is using Symfony as backend and Angular as frontend (I see you coming! the same issue will occur with react).

You run Symfony with the following command:

```
symfony local:server:start # aliased as symfony serve
```

And Angular with:

```
ng serve
```

Well configured, Symfony will serve your website on `https://you.wip` and Angular will serve `http://127.0.0.1:4200`.

This is an issue because it does not reflect what will happen in production most part of the time:
- Not the same domain
- Not the same port
- Not the same protocol

It may seems ok to you but if your production exposes the API on /api and the front on the same domain, it means you need to setup CORS only for development environment. It has other problems because as today many features require https in frontend, same domain. (I think of webrtc, popups, iframes, cookies...)

Symfony-cli is super limited so I think using it is a dead-end for this case (which happens for barely all projects I work on, am I the only one ?!).

And I'm pretty sure most people will use `docker`, `mkcert` and a router such as `caddy` (recommended by API platform team) or `traefik` with hardcoded hosts. I see 2 problems here:
- Super slow on macos (symfony-cli was a real help here!)
- Does not work on Windows (even with WSL2 you will experiment serious issues due to the super-custom configuration Microsoft did to the networking with WSL)
- Will not work out of the box because you need to modify your DNS
- For some reason (??) I experimented issues with mkcert certificates and not with symfony-cli ones
- The configuration with docker and everything is complex, you need people really competent with docker to maintain it, even if your project is not hosted with docker... This feels bad to me.

I think Symfony should provide a simple way to deal with all of this, and maybe I'm missing something...

So here I stand. Am I missing something or is it some work to do to make the whole thing simple and clean? Let's discuss it! Please add your point of view on the situation :) .

**Comment by LouisAugry** [comment, marked as answer]
Hi Nek!

That's a great question you're asking there, I'll be very curious to see the future answers.

I don't necessarily have any architecture to offer you, just some cheats.

You can serve angular over https - **SAME PROTOCOL**
```
ng serve --ssl
```

You can create a `proxy.conf.json` - **fake SAME DOMAIN and PORT**
```
{
    "/api": {
        "target": "https://you.wip",
        "secure": true
    }
}
```
then change your package.json to use this proxy
```
"scripts": {
    "start": "ng serve --proxy-config proxy.conf.json",
    ...
}
```

**Reply by Nek-** [reply]
Hello! Thank you very much it really looks exactly like what I want! Actually, it's perfect for my use case.

Too bad it's only for angular. But other frontend frameworks may also have tools like this.

Just one thing if somebody arrives on this post: it will not work with the domain `.wip` but the Symfony cli expose also by default on a port, and it works:

```json
{
    "/api": {
        "target": "https://127.0.0.1:8000",
        "secure": false
    }
}
```

_`secure` at false because it's self-signed, [learn more here](https://webpack.js.org/configuration/dev-server/#devserverproxy)._

To run my frontend server I run this command:

```
ng serve --proxy-config proxy.conf.json --ssl true --ssl-cert ssl/acme.local.pem --ssl-key ssl/acme.local-key.pem --public-host acme.local
```

And my README is the following:

```markdown
1. Install mkcert
2. Add the host `acme.local` to your local hosts file
3. Run `make setup`
```

Not a too complicated infrastructure, working on macos out of the box. Well, it's really good. Thanks again!

**Comment by ro0NL** [comment]
> Symfony will serve your website on https://you.wip and Angular will serve http://127.0.0.1:4200

Im not sure this is up to Symfony to solve. In that sense we need reproducible environments that _just work_ and can be copy/pasted to some degree. I concluded, today, all this is out-of-scope from regular application engineers :}

> I'll be very curious to see the future answers

Same.

**Reply by ro0NL** [reply]
To clarify, i was able to setup a decent/generic CI pipline using Github Actions in +- 2 weeks.

From what i know setting up "decent" DTAP from A to Z across many projects is very chaotic, which i cannot explain other than "random numbers" (that is ports&ips, rather than managed/automated **named** domains).

**Reply by Nek-** [reply]
I have no faith. I'm exposing a problem, a solution may come from the Symfony community or from others. As I stayed already, API Platform team seems to trust in Caddy: it has issues to me, but it's an answer.

@LouisAugry suggested ng-cli as an answer, and it's working great, but limited to Angular (but ok for me!).

**Comment by t0wqa** [comment]
Hi! 

I think you should abstract from what exact frameworks you use in backend or frontend. 

Your Symfony app shouldn't know who are the clients of its API, whether it is Angular or React or even Android/iOs app. 

And so should the frontend app: it shouldn't know, which exact languages/frameworks/infrastructure tools are used on API side. 

Separation of backend and frontend is a really good practice. 

And, well, actually, as for frontend and API domains in prod, usually they are different. For example, you may have site.com as front domain and api.site.com as API domain. 

So, from the Symfony side, you should just expose the API. If you need CORS (and, as you mentioned, in case of different domains you will definitely need it) you may use NelmioCorsBundle or similar solution, which allows you to centralize CORS management for API on the backend. 

From the Angular side, you should just consume the API using the backend domain. Locally it may be, for example, localhost:8080, on prod it would be api.site.com. And, important, your Angular shouldn't know, whether it is Symfony or Yii or Golang or anything else used in API.
